### DISCUSSION TITLE
Questions about implementing custom USB device function, callback is not working

### DISCUSSION THREAD

**Discussion by 2songh** [zephyrproject-rtos/zephyr#64425] [Q&A]
**Environment**
Zephyr Version: 3.4.99

**Describe the bug**
I implemented the USB callback in/out function by referring to the link.
* Link: https://docs.zephyrproject.org/latest/connectivity/usb/device/usb_device.html

*****

A folder was added under the zephyr/subsys/usb/device path.
Three files were added under the folder.
* CMakeLists.txt
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/71010d69-281c-49e2-be0e-cdf7cc6a8496)

*****

* Kconfig
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/fcaafe77-2d20-4b56-8c70-9254534fbf26)

*****

* core.c
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/2911f130-3b7d-418f-aff8-776a826abc57)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/858260c4-ab5d-42ae-af29-1120f68cad45)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/35f19ade-c100-4272-8407-9d63ad465fec)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/13b28f45-21d7-415f-a4af-f3a25deefb1b)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/91aa0940-4173-4d3a-aa14-14e705de6482)

*****

After that, I added USB-related configs to my project prj.conf file.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/4a87a003-44bc-43bb-84a0-57e9083dc6c6)

*****

When I built and flashed the project, USB worked fine.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/1fc5b8ff-32ef-473b-8a01-c3a78576aac7)

When connected to a PC, a message is displayed including the configuration I added.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/245d0803-6278-4f35-97c4-f3f69c1a5fb7)

However, even if I try to generate an interrupt by sending a message to my board through the PC, no callback function is generated.

Please advise if I have made any mistakes or are lacking in this matter.

**Comment by jfischer-no** [comment]
You need to use usb_write() to arm/write-to an endpoint and when the transfer to the host is complete, the driver will call endpoint callback m2i_in_cb.

**Reply by 2songh** [reply]
I'm already using usb_write() in m2i_in_cb.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/a9d2e01a-639b-4060-ac57-297d0bec647d)

Are there any problems in this part?

*****

Additionally, the dts file is missing from the main description, so it is added.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/ce1d391e-c631-4ead-8338-1436c9d6808a)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/126455296/ad15bc51-b030-45f9-8093-0dd6b7524316)
